productivity-suite/piercing-library/src/reframed-client.js

```diff
@@ -0,0 +1,241 @@
+  "getElementsByTagNameNS",
+];
+for (const queryProperty of domQueryProperties) {
+  document[queryProperty] = function reframedQueryFn() {
+    return reframedContainer[queryProperty].apply(reframedContainer, arguments);
+  };
+}
+
+// scope document.getElementById to execute just on the reframed container
+document.getElementById = function reframedGetElementById(id) {
+  return reframedContainer.querySelector(`#${id}`);
+};
+
+// scope document.getElementsByName to execute just on the reframed container
+document.getElementsByName = function reframedGetElementsByName(name) {
+  return reframedContainer.querySelectorAll(`[name="${name}"]`);
+};
+
+const domListenerProperties = ["addEventListener", "removeEventListener"];
+for (const listenerProperty of domListenerProperties) {
+  const originalDocumentFn = document[listenerProperty];
+  document[listenerProperty] = function reframedListenerFn(eventName) {
+    if (eventName === "DOMContentLoaded") {
+      return originalDocumentFn.apply(document, arguments);
+    }
+    return reframedContainer[listenerProperty].apply(
+      reframedContainer,
+      arguments
+    );
+  };
+
+  //const originalWindowFn = window[listenerProperty];
+  window[listenerProperty] = function reframedListenerFn(eventName) {
+    if (eventName === "popstate") {
+      // the popstate event fires only in the iframe
+      //Reflect.apply(originalWindowFn, window, arguments);
+    }
+    return reframedWindow[listenerProperty].apply(reframedWindow, arguments);
+  };
+}
+
+Object.defineProperty(document, "activeElement", {
+  get: () => {
+    return reframedDocument.activeElement;
+  },
+});
+
+const originalDocumentBody = document.body;
+Object.defineProperty(document, "body", {
+  get: () => {
+    return reframedContainer;
+  },
+});
+
+//const originalDocumentHead = document.head;
+Object.defineProperty(document, "head", {
+  get: () => {
+    return reframedContainer;
+  },
+});
+
+//const originalDocumentStylesheets = document.stylesheets;
+Object.defineProperty(document, "stylesheets", {
+  get: () => {
+    return reframedDocument.stylesheets;
+  },
+});
+
+// window.location is read-only and non-configurable, so we can't patch it
+//
+// additionally in a browsing context with one or more iframes, the history
+// all frames contribute to the joint history: https://www.w3.org/TR/2011/WD-html5-20110525/history.html#joint-session-history
+// this means that we need to be careful not to add duplicate entries to the
+// history stack via pushState within the iframe as that would double the
+// number of history entries that back/forward button would have to work through
+//
+// therefore we do the following:
+// - intercept all history.pushState history.replaceState calls and replay
+//   them in the main window
+// - update the window.location within the iframe via history.replaceState
+// - intercept window.addEventListener('popstate', ...) registration and forward it onto the main window
```

Comment:
I don't think we do anything like this in the demo? could we strip the history logic out so to make the code simpler? (and mention this possible implementation in an md file)

Reply:
@IgorMinar thoughts on this?

Reply:
Ok to strip for now